src/coreclr/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.Shared.xml

```diff
@@ -9,6 +9,9 @@
     <!-- Accessed via native code. -->
     <type fullname="System.Runtime.InteropServices.CustomMarshalers.*" />
 
+    <!-- Share definition across managed and native code. -->
```

Comment:
> I do not understand this comment. It should not be necessary to root the type just to share the layout with the native VM code.

The issue is the trimmer sees two fields that are unset and unused and therefore there are removed. They are set in unmanaged code, but the type needs to have the same layout in both. The layout doesn't need to be defined in `corelib.h` because it is directly defined in unmanaged, so `corelib.h` has no notion of `Signature`.

> Should this use corelib.h instead?

Perhaps, but it is going to introduce yet another description of the type's fields - managed, unmanaged and `corelib.h`. The definitions in `corelib.h` will never be used. And of course if someone does remove them then things break. I prefer to make the retention in the C# or XML side because that is clear from a trimmer case. The `corelib.h` file is simply a crutch.

Reply:
This is covered in https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/botr/corelib.md#types-with-a-managedunmanaged-duality. `DEFINE_CLASS_U` and `DEFINE_FIELD_U` macros in `corelib.h` are meant to be used for situations where we have managed and unmanaged layouts that are expected to match. Checked builds will assert when the type size or the field offsets do not match between the two layouts.

The [CreateRuntimeRootDescriptorFile task](https://github.com/dotnet/runtime/blob/3aa1ec5bb1f50f0a1bed9cfcac8734f742bcf24b/src/tools/illink/src/ILLink.Tasks/CreateRuntimeRootDescriptorFile.cs#L223-L231) understands `DEFINE_CLASS_U` and preserves all fields for types with managed/unmanaged duality.